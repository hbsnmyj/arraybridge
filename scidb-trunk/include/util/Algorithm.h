/*
**
* BEGIN_COPYRIGHT
*
* Copyright (C) 2008-2015 SciDB, Inc.
* All Rights Reserved.
*
* SciDB is free software: you can redistribute it and/or modify
* it under the terms of the AFFERO GNU General Public License as published by
* the Free Software Foundation.
*
* SciDB is distributed "AS-IS" AND WITHOUT ANY WARRANTY OF ANY KIND,
* INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY,
* NON-INFRINGEMENT, OR FITNESS FOR A PARTICULAR PURPOSE. See
* the AFFERO GNU General Public License for the complete license terms.
*
* You should have received a copy of the AFFERO GNU General Public License
* along with SciDB.  If not, see <http://www.gnu.org/licenses/agpl-3.0.html>
*
* END_COPYRIGHT
*/

#ifndef UTIL_ALGORITHM_H_
#define UTIL_ALGORITHM_H_

/****************************************************************************/
namespace scidb {
/****************************************************************************/
/**
 * A 3-sequence version of the std::transform() algorithm.
 */
template<class I,class J,class K,class out,class function>
out transform(I i,I e,J j,K k,out o,function f)
{
    for (; i!=e; ++i,++j,++k,++o)
    {
        *o = f(*i,*j,*k);
    }

    return o;
}

/**
 * A 4-sequence version of the std::transform() algorithm.
 */
template<class I,class J,class K,class L,class out,class function>
out transform(I i,I e,J j,K k,L l,out o,function f)
{
    for ( ; i!=e; ++i,++j,++k,++l,++o)
    {
        *o = f(*i,*j,*k,*l);
    }

    return o;
}

/****************************************************************************/
}
/****************************************************************************/
#endif
/****************************************************************************/
