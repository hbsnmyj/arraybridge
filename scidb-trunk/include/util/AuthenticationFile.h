/*
**
* BEGIN_COPYRIGHT
*
* Copyright (C) 2008-2015 SciDB, Inc.
* All Rights Reserved.
*
* SciDB is free software: you can redistribute it and/or modify
* it under the terms of the AFFERO GNU General Public License as published by
* the Free Software Foundation.
*
* SciDB is distributed "AS-IS" AND WITHOUT ANY WARRANTY OF ANY KIND,
* INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY,
* NON-INFRINGEMENT, OR FITNESS FOR A PARTICULAR PURPOSE. See
* the AFFERO GNU General Public License for the complete license terms.
*
* You should have received a copy of the AFFERO GNU General Public License
* along with SciDB.  If not, see <http://www.gnu.org/licenses/agpl-3.0.html>
*
* END_COPYRIGHT
*/


/**
 * @file AuthenticationFile.h
 *
 * @brief Utility for retrieving authentication information from an authentication file.
 *
 * @author Marty Corbett <mcorbett@paradigm4.com>
 */

#ifndef AUTHENTICATION_FILE_H_
#define AUTHENTICATION_FILE_H_

#include <stdlib.h>
#include <boost/property_tree/ptree.hpp>
#include <boost/property_tree/ini_parser.hpp>
#include <boost/regex.hpp>
#include <string>

namespace scidb
{
    /**
     * A class for accessing an authentication file similiar in structure to the config.ini file.
     * Currently this file will contain the username and password.  But may grow to include other
     * authentication specific settings.
     */
    class AuthenticationFile
    {
    private:
        boost::property_tree::ptree     _pt;

        bool checkAgainstRegularExpression(const std::string &str) const
        {
            static boost::regex  re("[A-Za-z0-9,-_./]+");
            boost::cmatch what;

            if(boost::regex_match(str.c_str(), what, re))
            {
                return true;
            }

            return false;
        }

    public:
        AuthenticationFile(const std::string &fileName)
        {
            boost::property_tree::ini_parser::read_ini(fileName.c_str(), _pt);
        }

        const std::string getUserName() const
        {
            std::string name = _pt.get<std::string>("security_password.user-name");
            if(false == checkAgainstRegularExpression(name))
            {
                name = "";
            }

            return name;
        }

        const std::string getUserPassword() const
        {
            std::string password = _pt.get<std::string>("security_password.user-password");
            if(false == checkAgainstRegularExpression(password))
            {
                password = "";
            }

            return password;
        }
    };
} // namespace scidb
#endif /* AUTHENTICATION_FILE_H_ */
