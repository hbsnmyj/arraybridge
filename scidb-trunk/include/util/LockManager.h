/*
**
* BEGIN_COPYRIGHT
*
* Copyright (C) 2008-2015 SciDB, Inc.
* All Rights Reserved.
*
* SciDB is free software: you can redistribute it and/or modify
* it under the terms of the AFFERO GNU General Public License as published by
* the Free Software Foundation.
*
* SciDB is distributed "AS-IS" AND WITHOUT ANY WARRANTY OF ANY KIND,
* INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY,
* NON-INFRINGEMENT, OR FITNESS FOR A PARTICULAR PURPOSE. See
* the AFFERO GNU General Public License for the complete license terms.
*
* You should have received a copy of the AFFERO GNU General Public License
* along with SciDB.  If not, see <http://www.gnu.org/licenses/agpl-3.0.html>
*
* END_COPYRIGHT
*/

/**
 * @file RWLock.h
 *
 * @author knizhnik@garret.ru
 *
 * @brief Read-Write lock implementation.
 *
 * Allow to lock in one thread and unlock in another.
 *
 */

#ifndef LOCK_MANAGER_H_
#define LOCK_MANAGER_H_

#include <pthread.h>
#include <map>
#include <string>

#include <system/Exceptions.h>
#include <util/RWLock.h>
#include <util/Singleton.h>

namespace scidb
{

class LockManager : public Singleton<LockManager>
{
private:
    Mutex mutex;
    std::map<std::string, std::shared_ptr<RWLock> > locks;

public:
    std::shared_ptr<RWLock> getLock(std::string const& arrayName)
    {
        ScopedMutexLock cs(mutex);
        std::shared_ptr<RWLock> lock = locks[arrayName];
        if (!lock) {
            locks[arrayName] = lock = std::shared_ptr<RWLock>(new RWLock());
        }
        return lock;
    }
};

} // namespace
#endif
