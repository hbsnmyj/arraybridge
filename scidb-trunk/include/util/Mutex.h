/*
**
* BEGIN_COPYRIGHT
*
* Copyright (C) 2008-2015 SciDB, Inc.
* All Rights Reserved.
*
* SciDB is free software: you can redistribute it and/or modify
* it under the terms of the AFFERO GNU General Public License as published by
* the Free Software Foundation.
*
* SciDB is distributed "AS-IS" AND WITHOUT ANY WARRANTY OF ANY KIND,
* INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY,
* NON-INFRINGEMENT, OR FITNESS FOR A PARTICULAR PURPOSE. See
* the AFFERO GNU General Public License for the complete license terms.
*
* You should have received a copy of the AFFERO GNU General Public License
* along with SciDB.  If not, see <http://www.gnu.org/licenses/agpl-3.0.html>
*
* END_COPYRIGHT
*/

/**
 * @file Mutex.h
 *
 * @author roman.simakov@gmail.com
 *
 * @brief The Mutex class for synchronization
 */

#ifndef MUTEX_H_
#define MUTEX_H_

#include <assert.h>
#include <errno.h>
#include <stdlib.h>
#include <pthread.h>

#include <sys/time.h>             // linux specific
#include <sys/resource.h>         // linux specific

#include <util/Platform.h>
#include <util/PerfTime.h>

#include <system/Exceptions.h>


namespace scidb
{

class Event;

class Mutex
{
friend class Event;
friend class Semaphore;
private:
 class PAttrEraser
 {
 public:
 PAttrEraser(pthread_mutexattr_t *attrPtr) : _attrPtr(attrPtr)
     {
         assert(_attrPtr!=NULL);
     }
     ~PAttrEraser()
     {
         pthread_mutexattr_destroy(_attrPtr);
     }
 private:
     pthread_mutexattr_t *_attrPtr;
 };

    pthread_mutex_t _mutex;
    static const int _mutexType = PTHREAD_MUTEX_RECURSIVE;

  public:
    void checkForDeadlock() {
        assert(_mutex.__data.__count == 1);
    }

    Mutex()
    {
        pthread_mutexattr_t mutexAttr;
        if (pthread_mutexattr_init(&mutexAttr)) {
            throw SYSTEM_EXCEPTION(SCIDB_SE_INTERNAL, SCIDB_LE_OPERATION_FAILED) << "pthread_mutexattr_init";
        }
        PAttrEraser onStack(&mutexAttr);

        if (pthread_mutexattr_settype(&mutexAttr, _mutexType)) {
            throw SYSTEM_EXCEPTION(SCIDB_SE_INTERNAL, SCIDB_LE_OPERATION_FAILED) << "pthread_mutexattr_settype";
        }
        if (pthread_mutex_init(&_mutex, &mutexAttr)) {
            throw SYSTEM_EXCEPTION(SCIDB_SE_INTERNAL, SCIDB_LE_OPERATION_FAILED) << "pthread_mutex_init";
        }
    }

    ~Mutex()
    {
        if (pthread_mutex_destroy(&_mutex)) {
            throw SYSTEM_EXCEPTION(SCIDB_SE_INTERNAL, SCIDB_LE_OPERATION_FAILED) << "pthread_mutex_destroy";
        }
    }

    void lock(scidb::perfTimeCategory_t tc = PTCW_MUT_OTHER)
    {
        {
            ScopedWaitTimer timer(tc);     // destruction updates the timing of tc
            if (pthread_mutex_lock(&_mutex)) {
                throw SYSTEM_EXCEPTION(SCIDB_SE_INTERNAL, SCIDB_LE_OPERATION_FAILED) << "pthread_mutex_lock";
            }
        }
    }

    void unlock()
    {
        if (pthread_mutex_unlock(&_mutex)) {
            throw SYSTEM_EXCEPTION(SCIDB_SE_INTERNAL, SCIDB_LE_OPERATION_FAILED) << "pthread_mutex_unlock";
        }
    }

    /// @return true if the mutex is locked by this thread; otherwise, false
    /// @note Works only in DEBUG mode
    /// @note Specific to Linux implementation of pthreads
    bool isLockedByThisThread()
    {
        bool result = false;
        if (isDebug())  {
            assert(_mutexType == PTHREAD_MUTEX_RECURSIVE);

            int locked = pthread_mutex_trylock(&_mutex);
            if (locked == 0) {
                result = (_mutex.__data.__count > 1);
                unlock();
            }
            result = result || (locked == EAGAIN) || (locked == EDEADLK);
        }
        return result;
    }
};


/***
 * RAII class for holding Mutex in object visible scope.
 */
class ScopedMutexLock
{
private:
	Mutex& _mutex;

public:
	ScopedMutexLock(Mutex& mutex, perfTimeCategory_t tc = PTCW_SEM_OTHER): _mutex(mutex)
	{
		_mutex.lock(tc);
	}

	~ScopedMutexLock()
	{
		_mutex.unlock();
	}
};


} //namespace

#endif /* MUTEX_H_ */
