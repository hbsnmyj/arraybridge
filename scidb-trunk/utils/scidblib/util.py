#!/usr/bin/python

# BEGIN_COPYRIGHT
#
# Copyright (C) 2008-2015 SciDB, Inc.
# All Rights Reserved.
#
# SciDB is free software: you can redistribute it and/or modify
# it under the terms of the AFFERO GNU General Public License as published by
# the Free Software Foundation.
#
# SciDB is distributed "AS-IS" AND WITHOUT ANY WARRANTY OF ANY KIND,
# INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY,
# NON-INFRINGEMENT, OR FITNESS FOR A PARTICULAR PURPOSE. See
# the AFFERO GNU General Public License for the complete license terms.
#
# You should have received a copy of the AFFERO GNU General Public License
# along with SciDB.  If not, see <http://www.gnu.org/licenses/agpl-3.0.html>
#
# END_COPYRIGHT

# BEGIN_COPYRIGHT_EXCEPTIONS
#
# All code in this file is covered by the copyright detailed above,
# with the following explicit exceptions.  See doc strings for
# attribution and/or copyright information.
#
#   - superTuple
#
# END_COPYRIGHT_EXCEPTIONS

import getpass
import os
import subprocess
from collections import namedtuple
from operator import itemgetter

def superTuple(typename, *attribute_names):
    """Create and return a subclass of 'tuple', with named attributes.

    THIS FUNCTION IS OBSOLETE!  USE collections.namedtuple INSTEAD!!!

    Sample Usage:
        Employee = superTuple('Employee', 'name', 'phone', 'email')
        joe = Employee('Joe Smith', '617-555-1212', 'jsmith@example.com')
        if joe.phone.startswith('617-'):
            boston_employees.append(joe)

    @see "Python Cookbook", Martelli et. al., O'Reilly, 2nd ed.,
    recipe 6.7 "Implementing Tuples with Named Items".
    """
    # Make the subclass with appropriate __new__ and __repr__ methods.
    nargs = len(attribute_names)
    class supertup(tuple):
        __slots__ = () # save memory, we don't need per-instance dict
        def __new__(cls, *args):
            if len(args) != nargs:
                raise TypeError("%s takes exactly %d arguments (%d given)" % (
                    typename, nargs, len(args)))
            return tuple.__new__(cls, args)
        def __repr__(self):
            return '%s(%s)' % (typename, ', '.join(map(repr, self)))
    # Set up attribute names for gettable indices.
    for i, attr_name in enumerate(attribute_names):
        setattr(supertup, attr_name, property(itemgetter(i)))
    supertup.__name__ = typename
    return supertup


def getVerifiedPassword(prompt=None, verify=None):
    """Read and verify a password from the tty.

    @param prompt the prompt string for initial password entry
    @param verify the prompt string for verification
    """
    if prompt is None:
        prompt = "Password: "
    if verify is None:
        verify = "Re-enter password: "
    while True:
        p1 = getpass.getpass(prompt)
        p2 = getpass.getpass(verify)
        if p1 == p2:
            break
        try:
            with open("/dev/tty", "w") as F:
                print >>F, "Passwords do not match"
        except OSError:
            print >>sys.stderr, "Passwords do not match"
    return p1


def iquery(*args, **kwargs):
    """Run an iquery command.  First argument should be '-aq' or similar.

    Unfortunately the code in scidb_afl.py is very unpythonic, hence
    this minimalist iquery command wrapper.  SciDB host and port are
    taken from keyword args or from the environment in that order.

    Keyword Args:
    host - SciDB cluster host (e.g. $IQUERY_HOST, default 'localhost').
    port - SciDB query port (e.g. $IQUERY_PORT, default 1239).
    """
    cmd = ['iquery']
    if 'host' in kwargs:
        cmd.extend(['-c', kwargs['host']])
    elif 'IQUERY_HOST' in os.environ:
        cmd.extend(['-c', os.environ['IQUERY_HOST']])
    if 'port' in kwargs:
        cmd.extend(['-p', str(kwargs['port'])])
    elif 'IQUERY_PORT' in os.environ:
        cmd.extend(['-p', os.environ['IQUERY_PORT']])
    cmd.extend(map(str, args))
    p = subprocess.Popen(cmd,
                         stdout=subprocess.PIPE,
                         stderr=subprocess.STDOUT)
    return p.communicate()[0]


def make_table(entry_name, query, **kwargs):
    """Build a list of named tuples based on result of the given AFL query.

    @param entry_name name of type to be created by collections.namedtuple
    @param query      AFL query from whose output we will make a table
    @param kwargs     keyword args for iquery() call

    Because the entire query result is read into memory, best to use
    this only with queries returning smallish results.

    All returned namedtuple fields are strings; we do not (yet)
    automatically convert AFL data types to their Python equivalents.

    An example:
    >>> t = make_table('ArrayTable', "list('arrays',true)")
    >>> all_versioned_array_ids = [x.aid for x in t if x.aid != x.uaid]
    """
    # Format tsv+:l gives dimension/attribute names for use as tuple attributes.
    table_data = iquery('-otsv+:l', '-aq', query, kwargs).splitlines()
    if "Error id:" in table_data:
        raise RuntimeError(table_data)
    # Sometimes SciDB can give the same label to >1 attribute; make them unique.
    attrs = []
    seen = dict()
    for label in table_data[0].split():
        if label in seen:
            seen[label] += 1
            label = '_'.join((label, str(seen[label])))
        else:
            seen[label] = 1
        attrs.append(label)
    # Create our data type and fill in the table.
    tuple_type = namedtuple(entry_name, attrs)
    table = []
    for line in table_data[1:]:
        table.append(tuple_type._make(line.split('\t')))
    return table
