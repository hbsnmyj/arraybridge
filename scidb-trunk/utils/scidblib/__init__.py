########################################
# BEGIN_COPYRIGHT
#
# Copyright (C) 2008-2015 SciDB, Inc.
# All Rights Reserved.
#
# SciDB is free software: you can redistribute it and/or modify
# it under the terms of the AFFERO GNU General Public License as published by
# the Free Software Foundation.
#
# SciDB is distributed "AS-IS" AND WITHOUT ANY WARRANTY OF ANY KIND,
# INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY,
# NON-INFRINGEMENT, OR FITNESS FOR A PARTICULAR PURPOSE. See
# the AFFERO GNU General Public License for the complete license terms.
#
# You should have received a copy of the AFFERO GNU General Public License
# along with SciDB.  If not, see <http://www.gnu.org/licenses/agpl-3.0.html>
#
# END_COPYRIGHT
########################################

"""This directory contains some common Python modules shared by SciDB's scripts."""

class AppError(Exception):
    """An exception class that multiple modules in scidblib may use."""
    pass

import counter
import pgpass_updater
import util
import scidb_afl as afl
import scidb_control as control
import scidb_math as math
import scidb_progress as progress
import scidb_psf as psf
import scidb_schema as schema
import statistics
