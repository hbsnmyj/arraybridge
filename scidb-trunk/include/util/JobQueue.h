/*
**
* BEGIN_COPYRIGHT
*
* Copyright (C) 2008-2015 SciDB, Inc.
* All Rights Reserved.
*
* SciDB is free software: you can redistribute it and/or modify
* it under the terms of the AFFERO GNU General Public License as published by
* the Free Software Foundation.
*
* SciDB is distributed "AS-IS" AND WITHOUT ANY WARRANTY OF ANY KIND,
* INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY,
* NON-INFRINGEMENT, OR FITNESS FOR A PARTICULAR PURPOSE. See
* the AFFERO GNU General Public License for the complete license terms.
*
* You should have received a copy of the AFFERO GNU General Public License
* along with SciDB.  If not, see <http://www.gnu.org/licenses/agpl-3.0.html>
*
* END_COPYRIGHT
*/

/**
 * @file JobQueue.h
 *
 * @author roman.simakov@gmail.com
 *
 * @brief The queue of jobs for execution in thread pool
 */

#ifndef JOBQUEUE_H_
#define JOBQUEUE_H_

#include <list>
#include <memory>

#include <util/Job.h>
#include <util/Mutex.h>
#include <util/Semaphore.h>
#include <util/RWLock.h>

namespace scidb
{

class JobQueue
{
private:
    std::list< std::shared_ptr<Job> > _queue;
    Mutex _queueMutex;
    Semaphore _queueSemaphore;

public:
    JobQueue();

    size_t getSize() const
    {
        return _queue.size();
    }

    /// Add new job to the end of queue
    void pushJob(std::shared_ptr<Job> job);

    /// Add new job to the beginning of queue
    void pushHighPriorityJob(std::shared_ptr<Job> job);

    /**
     * Get next job from the beginning of the queue
     * If there is next element the method waits
     */
    std::shared_ptr<Job> popJob();
};

} // namespace

#endif /* JOBQUEUE_H_ */
