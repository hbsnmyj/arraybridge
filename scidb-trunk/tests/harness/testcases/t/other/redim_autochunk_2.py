#!/usr/bin/python

"""Some basic functional tests for autochunking redimension()."""

import argparse
import os
import sys
import traceback

from scidblib.util import (iquery, make_table)
from box_of_points import main as boxofpoints
from collections import defaultdict

_args = None                    # Globally visible parsed arguments
_pgm = None                     # Globally visible program name
_tmpdir = None                  # Dir for generated data files

class AppError(Exception):
    """Base class for all exceptions that halt script execution."""
    pass

def dbg(*args):
    #print >>sys.stderr, "DBG:", ' '.join(map(str, args))
    print "DBG:", ' '.join(map(str, args))

def noerr(s):
    """Predicate for easy testing for SciDB errors."""
    return "Error id:" not in s

class TheTest(object):

    """
    A dirt-simple test runner.

    I wanted to use unittest for this, but the version in Python 2.6
    is so old and crufty that it doesn't support setUpClass() and
    setUpModule() methods.  We should install "nose" everywhere, IMHO
    it's much better than unittest. -mjl
    """

    def setUpClass(self):
        """Create some test data files used by all test methods."""
        print "Setup ...",
        sys.stdout.flush()
        self._array_cleanups = []
        self._files = {}        # map array name to input file
        # Put all our data files in one temp directory so we can
        # easily remove them all during tearDown.
        if os.system("rm -rf {0} ; mkdir -p {0}".format(_tmpdir)):
            raise AppError("Trouble (re)creating %s" % _tmpdir)

        # Create slightly sparse 3-D input data with no collisions.
        self._files['nocoll_3d'] = os.path.join(_tmpdir, "nocoll_3d.bin")
        if boxofpoints(['boxofpoints',
                        '--lower-corner', '0,0,0',
                        '--upper-corner', '9,99,99',
                        '--cells', '80000', # sparse: 80,000 < 10x100x100
                        '--format', 'binary',
                        '--output', self._files['nocoll_3d'],
                        '--seed', '42']):
            raise AppError("box_of_points could not create %s" %
                           self._files['nocoll_3d'])

        # Create dense 2-D input data with 10% collisions.
        self._files['coll_2d'] = os.path.join(_tmpdir, "coll_2d.bin")
        if boxofpoints(['boxofpoints',
                        '--lower-corner', '0,0',
                        '--upper-corner', '99,999',
                        '--cells', '100000', # dense: 100,000 == 100x1000
                        '--collisions', '0.1', # 10% collision rate
                        '--format', 'binary',
                        '--output', self._files['coll_2d'],
                        '--seed', '42']):
            raise AppError("box_of_points could not create %s" %
                           self._files['coll_2d'])
        print "done"

    def tearDownClass(self):
        print "Teardown ...",
        sys.stdout.flush()
        if not _args.keep_arrays:
            if os.system("rm -rf {0}".format(_tmpdir)):
                raise AppError("Trouble cleaning up %s" % _tmpdir)
            for a in self._array_cleanups:
                iquery('-naq', "remove(%s)" % a)
        print "done"

    def test_00_load_3d_ac(self):
        """Load 3-D no-collision data using autochunking"""
        dims = "x,y,z"          # Autochunked!
        query = """
            store(
              redimension(
                input(<v:int64,x:int64,y:int64,z:int64>[dummy], '{0}',
                      -2, '(int64,int64,int64,int64)'),
                <v:int64>[{1}]),
              {2}) """.format(self._files['nocoll_3d'], dims, "nocoll_3d_ac")
        out = iquery('-naq', query)
        assert noerr(out), out
        self._array_cleanups.append("nocoll_3d_ac")

    def test_01_load_3d_concrete(self):
        """Load 3-D no-collisions data with specified chunks"""
        dims = "x=0:*,10,0,y=0:*,100,0,z=0:*,100,0"
        query = """
            store(
              redimension(
                input(<v:int64,x:int64,y:int64,z:int64>[dummy], '{0}',
                      -2, '(int64,int64,int64,int64)'),
                <v:int64>[{1}]),
              {2}) """.format(self._files['nocoll_3d'], dims, "nocoll_3d")
        out = iquery('-naq', query)
        assert noerr(out), out
        self._array_cleanups.append("nocoll_3d")

    def test_02_nocoll_3d_counts_and_sums(self):
        """Compare 3-D array counts and sums"""
        out1 = iquery('-otsv', '-aq', 'aggregate(nocoll_3d_ac,count(*),sum(v))')
        assert noerr(out1), out1
        out2 = iquery('-otsv', '-aq', 'aggregate(nocoll_3d,count(*),sum(v))')
        assert noerr(out2), out2
        c1, s1 = map(int, out1.split())
        c2, s2 = map(int, out2.split())
        assert c1 == c2, "counts differ"
        assert s1 == s2, "sums differ"

    def test_03_nocoll_3d_check_values(self):
        """Cell-by-cell value comparison for 3-D arrays"""
        out = iquery('-otsv+', '-aq', """
            filter(join(nocoll_3d,nocoll_3d_ac), nocoll_3d.v <> nocoll_3d_ac.v)
            """)
        assert noerr(out), out
        assert out == '', "Cell values differ:\n\t{0}".format(out)

    def test_04_load_2d_ac_w_collisions(self):
        """Load 2-D data containing collisions using autochunking"""
        dims = "x=0:*,*,0,y=0:*,*,0,synth=0:*,*,0"
        query = """
            store(
              redimension(
                input(<v:int64,x:int64,y:int64>[dummy], '{0}',
                      -2, '(int64,int64,int64)'),
                <v:int64>[{1}]),
              {2}) """.format(self._files['coll_2d'], dims, "coll_2d_ac")
        out = iquery('-naq', query)
        assert noerr(out), out
        self._array_cleanups.append("coll_2d_ac")

    def test_05_load_2d_concrete_w_collisions(self):
        """Load 2-D data containing collisions with specified chunks"""
        dims = "x=0:*,100,0,y=0:*,100,0,synth=0:9,10,0"
        query = """
            store(
              redimension(
                input(<v:int64,x:int64,y:int64>[dummy], '{0}',
                      -2, '(int64,int64,int64)'),
                <v:int64>[{1}]),
              {2}) """.format(self._files['coll_2d'], dims, "coll_2d")
        out = iquery('-naq', query)
        assert noerr(out), out
        self._array_cleanups.append("coll_2d")

    def test_06_coll_2d_counts_and_sums(self):
        """Compare 2-D array counts and sums"""
        out1 = iquery('-otsv', '-aq', 'aggregate(coll_2d_ac,count(*),sum(v))')
        assert noerr(out1), out1
        out2 = iquery('-otsv', '-aq', 'aggregate(coll_2d,count(*),sum(v))')
        assert noerr(out2), out2
        c1, s1 = map(int, out1.split())
        c2, s2 = map(int, out2.split())
        assert c1 == c2, "counts differ"
        assert s1 == s2, "sums differ"

    def test_07_coll_2d_check_values(self):
        """Cell-by-cell value comparison for 2-D arrays

        This test is complicated by the fact that with different chunk
        intervals, redimension() does not produce synthetic dimension
        siblings in any particular order.  So we must process the
        filtered list of differing cells, and only complain if the
        *set* of values along the synthetic dimension at [x,y,*]
        differs for the two arrays.  For example, if the two arrays
        held

            {x,y,synth} v           {x,y,synth} v
            {2,7,0} 20              {2,7,0} 20
            {2,7,1} 73              {2,7,1} 99
            {2,7,2} 99              {2,7,2} 73

        that is perfectly fine.
        """
        tbl = make_table('CellDiffs', """
            filter(join(coll_2d,coll_2d_ac), coll_2d.v <> coll_2d_ac.v)
            """)
        v_xy_sets = defaultdict(set)
        v2_xy_sets = defaultdict(set)
        for celldiff in tbl:
            key = (int(celldiff.x), int(celldiff.y))
            v_xy_sets[key].add(int(celldiff.v))
            v2_xy_sets[key].add(int(celldiff.v_2))
        assert len(v_xy_sets) == len(v2_xy_sets)
        for xy in v_xy_sets:
            assert v_xy_sets[xy] == v2_xy_sets[xy], \
                "Synthetic dimension trouble at {0}".format(xy)

    def test_08_load_3d_ac_w_overlap(self):
        """Load 3-D no-collision data using autochunking and overlaps."""
        dims = "x=0:*,*,2;y=0:*,*,3;z=0:*,*,0" # Autochunked!
        query = """
            store(
              redimension(
                input(<v:int64,x:int64,y:int64,z:int64>[dummy], '{0}',
                      -2, '(int64,int64,int64,int64)'),
                <v:int64>[{1}]),
              {2}) """.format(self._files['nocoll_3d'], dims, "nocoll_3d_ac_ol")
        out = iquery('-naq', query)
        assert noerr(out), out
        self._array_cleanups.append("nocoll_3d_ac_ol")

    def test_09_nocoll_3d_overlap_counts_and_sums(self):
        """Compare 3-D array counts and sums (overlap)"""
        out1 = iquery('-otsv', '-aq',
                      'aggregate(nocoll_3d_ac_ol,count(*),sum(v))')
        assert noerr(out1), out1
        out2 = iquery('-otsv', '-aq', 'aggregate(nocoll_3d,count(*),sum(v))')
        assert noerr(out2), out2
        c1, s1 = map(int, out1.split())
        c2, s2 = map(int, out2.split())
        assert c1 == c2, "counts differ"
        assert s1 == s2, "sums differ"

    def test_10_nocoll_3d_overlap_check_values(self):
        """Cell-by-cell value comparison for 3-D arrays (overlap)"""
        out = iquery('-otsv+', '-aq', """
            filter(join(nocoll_3d, nocoll_3d_ac_ol),
                        nocoll_3d.v <> nocoll_3d_ac_ol.v)
            """)
        assert noerr(out), out
        assert out == '', "Cell values differ:\n\t{0}".format(out)

    def test_11_empty_input(self):
        """Autochunked redimension of empty array should not fail (SDB-5109)"""
        out = iquery('-aq', 'create temp array empty<val:double>[k=0:39,20,4]')
        assert noerr(out), out
        self._array_cleanups.append("empty")
        out = iquery('-otsv+', '-aq',
                     'redimension(empty, <val:double>[k=0:39,*,3])')
        assert noerr(out), out
        assert not out, "Redim of empty array is not empty: '%s'" % out

    def test_12_one_input_cell(self):
        """Autochunked redimension of 1-cell array should not fail"""
        out = iquery('-aq', 'create temp array ONE<val:double>[k=0:39,20,4]')
        assert noerr(out), out
        self._array_cleanups.append("ONE")
        # Insert one cell at k == 25.
        out = iquery('-otsv+:l', '-naq', """
            insert(
              redimension(
                apply(build(<val:double>[i=0:0,1,0], 3.14), k, 25),
                ONE),
            ONE)""")
        assert noerr(out), out
        out = iquery ('-otsv+', '-aq',
                      'redimension(ONE, <val:double>[k=0:39,*,3])')
        assert noerr(out), out
        try:
            numbers = map(float, out.split())
        except ValueError:
            assert False, "Unexpected non-number in '%s'" % out
        assert len(numbers) == 2
        assert numbers[0] == 25
        assert numbers[1] == 3.14

    def run(self):
        """Dirt-simple test runner."""
        self.setUpClass()
        tests = [x for x in dir(self) if x.startswith("test_")]
        tests.sort()
        errors = 0
        for t in tests:
            print t, "...",
            sys.stdout.flush()
            try:
                getattr(self,t)()
            except AssertionError as e:
                print "FAIL (%s)" % e
                errors += 1
            else:
                print "ok"
        self.tearDownClass()
        if errors:
            print "Errors in", errors, "of", len(tests), "tests."
        else:
            print "All", len(tests), "tests passed."
        return errors


def set_iquery_envariables():
    """Make sure scidblib.util.iquery() knows how to contact SciDB.

    Command line options override environment.  Environment overrides
    hard-coded defaults.
    """
    if _args.host:
        os.environ['IQUERY_HOST'] = _args.host
    elif 'IQUERY_HOST' not in os.environ:
        os.environ['IQUERY_HOST'] = 'localhost'
    if _args.port:
        os.environ['IQUERY_PORT'] = _args.port
    elif 'IQUERY_PORT' not in os.environ:
        os.environ['IQUERY_PORT'] = '1239'


def main(argv=None):
    """Argument parsing and last-ditch exception handling.

    See http://www.artima.com/weblogs/viewpost.jsp?thread=4829
    """
    if argv is None:
        argv = sys.argv

    global _pgm
    _pgm = "%s:" % os.path.basename(argv[0]) # colon for easy use by print

    parser = argparse.ArgumentParser(
        description="This skeleton program does very little.",
        epilog='Type "pydoc %s" for more information.' % _pgm[:-1])
    parser.add_argument('-c', '--host', default=None,
                        help='Target host for iquery commands.')
    parser.add_argument('-p', '--port', default=None,
                        help='SciDB port on target host for iquery commands.')
    parser.add_argument('-r', '--run-id', type=int, default=0,
                        help='Unique run identifier.')
    parser.add_argument('-k', '--keep-arrays', action='store_true',
                        help='Do not remove test arrays during cleanup.')
    parser.add_argument('-v', '--verbosity', default=0, action='count',
                    help='Increase debug logging. 1=info, 2=debug, 3=debug+')

    global _args
    _args = parser.parse_args(argv[1:])

    global _tmpdir
    _tmpdir = "/tmp/redim_autochunk_2.{0}".format(_args.run_id)

    set_iquery_envariables()

    try:
        tt = TheTest()
        return tt.run()
    except AppError as e:
        print >>sys.stderr, _pgm, e
        return 1
    except Exception as e:
        print >>sys.stderr, _pgm, "Unhandled exception:", e
        traceback.print_exc()   # always want this for unexpected exceptions
        return 2

if __name__ == '__main__':
    sys.exit(main())
