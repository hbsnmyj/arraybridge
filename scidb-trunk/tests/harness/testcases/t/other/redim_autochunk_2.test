# Some basic functional tests.
--shell --store --command "PYTHONPATH=${SCIDB_INSTALL_PATH}/bin ${TESTDIR}/redim_autochunk_2.py -c ${IQUERY_HOST} -p ${IQUERY_PORT} --run-id=${HPID}"
