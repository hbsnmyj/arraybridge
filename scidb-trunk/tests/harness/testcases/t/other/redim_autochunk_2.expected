Setup ... done
test_00_load_3d_ac ... ok
test_01_load_3d_concrete ... ok
test_02_nocoll_3d_counts_and_sums ... ok
test_03_nocoll_3d_check_values ... ok
test_04_load_2d_ac_w_collisions ... ok
test_05_load_2d_concrete_w_collisions ... ok
test_06_coll_2d_counts_and_sums ... ok
test_07_coll_2d_check_values ... ok
test_08_load_3d_ac_w_overlap ... ok
test_09_nocoll_3d_overlap_counts_and_sums ... ok
test_10_nocoll_3d_overlap_check_values ... ok
test_11_empty_input ... ok
test_12_one_input_cell ... ok
Teardown ... done
All 13 tests passed.

